AX: To avoid confusing ATs, WebKit should avoid exposing listboxes as listboxes if they don't have any option descendants

twilco opened the pull request:
#### 98548924f48f6e5eb83ca120f38b8bcbe720c580
<pre>
AX: To avoid confusing ATs, WebKit should avoid exposing listboxes as listboxes if they don&apos;t have any option descendants
<a href="https://bugs.webkit.org/show_bug.cgi?id=292443">https://bugs.webkit.org/show_bug.cgi?id=292443</a>
<a href="https://rdar.apple.com/150533266">rdar://150533266</a>

Reviewed by Joshua Hoffman.

If a role=&quot;listbox&quot; has no options, we should just call it a group to avoid confusing assistive technologies who expect
listboxes to have a certain structure. Said confusion can have adverse side effects, like VoiceOver failing to navigate
content within an invalid listbox.

This commit also removes unused AXProperty::IsListBox.

* LayoutTests/accessibility/mac/invalid-list-box-expected.txt: Added.
* LayoutTests/accessibility/mac/invalid-list-box.html: Added.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::isValidListBox const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::isInvalidListBox const):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
(WebCore::AXCoreObject::rolePlatformString):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: <a href="https://commits.webkit.org/294459@main">https://commits.webkit.org/294459@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/452f4719f4f5d9557e53e57e2ca35d4cc61269ff

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101848 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21516 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11832 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107006 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52482 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103888 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21824 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30023 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77547 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34566 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104855 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16863 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91968 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57884 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16690 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9989 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51837 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86546 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10065 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109364 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28981 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21347 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86524 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29342 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88171 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86097 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21915 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30856 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8577 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23152 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28909 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34200 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28720 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32043 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30279 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -209,6 +210,37 @@ bool AXCoreObject::isTextControl() const
+            return true;
+        // Don't iterate forever in case someone added role="listbox" to some high-level element.
+        // If we haven't found an option after checking 200 objects, this probably isn't valid anyways.
+        if (iterations >= 200)
```
Should we consider a number a little higher than 200? Since this is a BFS, let's say we have a list box with 201 items, each wrapped with just one extra div (with a `option` child). Then that will return false right?

twilco replied:
Bumped to 250. I have a hard time believing there is an actual, valid listbox in the wild that hits either limit 🤔

hoffmanjoshua reviewed (approved):
This looks good to me, just have one question about the limit we've set on the iterations.

twilco committed:
AX: To avoid confusing ATs, WebKit should avoid exposing listboxes as listboxes if they don't have any option descendants
https://bugs.webkit.org/show_bug.cgi?id=292443
rdar://150533266

Reviewed by Joshua Hoffman.

If a role="listbox" has no options, we should just call it a group to avoid confusing assistive technologies who expect
listboxes to have a certain structure. Said confusion can have adverse side effects, like VoiceOver failing to navigate
content within an invalid listbox.

This commit also removes unused AXProperty::IsListBox.

* LayoutTests/accessibility/mac/invalid-list-box-expected.txt: Added.
* LayoutTests/accessibility/mac/invalid-list-box.html: Added.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::isValidListBox const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::isInvalidListBox const):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
(WebCore::AXCoreObject::rolePlatformString):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: https://commits.webkit.org/294459@main
